File: internal/a2ui/a2ui_test.go
package a2ui

import "testing"

func TestFromWidgetConfirmation(t *testing.T) {
	msgs, ok := FromWidget(map[string]any{
		"_kind":    "widget/confirmation",
		"title":    "Подтверждение возврата",
		"message":  "Оформить возврат по заказу 1055?",
		"order_id": "1055",
		"actions": []any{
			map[string]any{"id": "approve", "label": "Оформить возврат"},
			map[string]any{"id": "decline", "label": "Отмена"},
		},
	})
	if !ok {
		t.Fatal("confirmation widget should map")
	}
	if len(msgs) != 2 {
		t.Fatalf("want 2 messages (createSurface, updateComponents), got %d", len(msgs))
	}
	if msgs[0]["version"] != "v0.9" || msgs[0]["createSurface"] == nil {
		t.Errorf("msg0 must be a v0.9 createSurface, got %#v", msgs[0])
	}
	uc, _ := msgs[1]["updateComponents"].(map[string]any)
	if uc == nil {
		t.Fatalf("msg1 must be updateComponents, got %#v", msgs[1])
	}
	comps, _ := uc["components"].([]map[string]any)
	// Expect at least: root, message text, two buttons, two button labels.
	var buttons, actions int
	for _, c := range comps {
		if c["component"] == "Button" {
			buttons++
			if a, ok := c["action"].(map[string]any); ok {
				if _, ok := a["event"].(map[string]any); ok {
					actions++
				}
			}
		}
	}
	if buttons != 2 || actions != 2 {
		t.Errorf("want 2 buttons each with an action.event, got buttons=%d actions=%d", buttons, actions)
	}
}

func TestFromWidgetUnknownKind(t *testing.T) {
	if _, ok := FromWidget(map[string]any{"_kind": "widget/nope"}); ok {
		t.Error("unknown kind must return ok=false")
	}
}

// collectTexts returns all Text component texts from a FromWidget result.
func collectTexts(t *testing.T, msgs []map[string]any) []string {
	t.Helper()
	uc, _ := msgs[1]["updateComponents"].(map[string]any)
	comps, _ := uc["components"].([]map[string]any)
	var out []string
	for _, c := range comps {
		if c["component"] == "Text" {
			if s, ok := c["text"].(string); ok {
				out = append(out, s)
			}
		}
	}
	return out
}

func TestFromWidgetOrderIncludesLink(t *testing.T) {
	msgs, ok := FromWidget(map[string]any{
		"_kind": "widget/order",
		"title": "Заказ 1041",
		"order": map[string]any{
			"id": "1041", "item": "USB-C хаб", "status_label": "доставлен",
			"amount": 34.5, "currency": "EUR",
			"url": "https://shop.test/orders/1041",
		},
	})
	if !ok {
		t.Fatal("order widget should map")
	}
	want := "[Открыть карточку заказа →](https://shop.test/orders/1041)"
	for _, s := range collectTexts(t, msgs) {
		if s == want {
			return
		}
	}
	t.Errorf("no markdown link text %q in components: %v", want, collectTexts(t, msgs))
}

func TestFromWidgetOrderListRowsLinkTheNumber(t *testing.T) {
	msgs, ok := FromWidget(map[string]any{
		"_kind": "widget/order_list",
		"title": "Последние заказы: alice",
		"orders": []any{map[string]any{
			"id": "1041", "item": "USB-C хаб", "status_label": "доставлен",
			"amount": 34.5, "currency": "EUR", "created": "2026-06-10",
			"url": "https://shop.test/orders/1041",
		}},
	})
	if !ok {
		t.Fatal("order_list widget should map")
	}
	texts := collectTexts(t, msgs)
	for _, s := range texts {
		if s == "[#1041](https://shop.test/orders/1041)  USB-C хаб — доставлен (34.5 EUR, 2026-06-10)" {
			return
		}
	}
	t.Errorf("no linked row in components: %v", texts)
}

func TestFromWidgetOrderWithoutURLHasNoLink(t *testing.T) {
	msgs, ok := FromWidget(map[string]any{
		"_kind": "widget/order",
		"title": "Заказ 1041",
		"order": map[string]any{"id": "1041", "item": "USB-C хаб"},
	})
	if !ok {
		t.Fatal("order widget should map")
	}
	for _, s := range collectTexts(t, msgs) {
		if len(s) > 0 && s[0] == '[' {
			t.Errorf("unexpected link text %q for widget without url", s)
		}
	}
}

File: java/orchestrator/src/test/java/io/github/kmpavloff/a2ademo/orchestrator/a2ui/A2uiTest.java
package io.github.kmpavloff.a2ademo.orchestrator.a2ui;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/** Port of the Go internal/a2ui tests. */
class A2uiTest {

    @Test
    @SuppressWarnings("unchecked")
    void fromWidgetConfirmation() {
        List<Map<String, Object>> msgs = A2ui.fromWidget(Map.of(
                "_kind", "widget/confirmation",
                "title", "Подтверждение возврата",
                "message", "Оформить возврат по заказу 1055? (да/нет)",
                "order_id", "1055",
                "actions", List.of(
                        Map.of("id", "approve", "label", "Оформить возврат"),
                        Map.of("id", "decline", "label", "Отмена"))));
        assertNotNull(msgs, "confirmation widget should map");
        assertEquals(2, msgs.size(), "want createSurface + updateComponents");
        assertEquals("v0.9", msgs.get(0).get("version"));
        assertNotNull(msgs.get(0).get("createSurface"));

        Map<String, Object> uc = (Map<String, Object>) msgs.get(1).get("updateComponents");
        assertNotNull(uc);
        List<Map<String, Object>> comps = (List<Map<String, Object>>) uc.get("components");
        int buttons = 0;
        int actions = 0;
        String msgText = "";
        for (Map<String, Object> c : comps) {
            if ("Button".equals(c.get("component"))) {
                buttons++;
                if (c.get("action") instanceof Map<?, ?> a && a.get("event") instanceof Map<?, ?>) {
                    actions++;
                }
            }
            if ("msg".equals(c.get("id"))) {
                msgText = String.valueOf(c.get("text"));
            }
        }
        assertEquals(2, buttons);
        assertEquals(2, actions);
        assertEquals("Оформить возврат по заказу 1055?", msgText, "(да/нет) hint must be dropped");
    }

    @Test
    void fromWidgetUnknownKind() {
        assertNull(A2ui.fromWidget(Map.of("_kind", "widget/nope")));
    }

    @SuppressWarnings("unchecked")
    private static List<String> texts(List<Map<String, Object>> msgs) {
        Map<String, Object> uc = (Map<String, Object>) msgs.get(1).get("updateComponents");
        return ((List<Map<String, Object>>) uc.get("components")).stream()
                .filter(c -> "Text".equals(c.get("component")))
                .map(c -> String.valueOf(c.get("text")))
                .toList();
    }

    @Test
    void orderWidgetWithUrlGetsMarkdownLink() {
        List<Map<String, Object>> msgs = A2ui.fromWidget(Map.of(
                "_kind", "widget/order",
                "title", "Заказ 1041",
                "order", Map.of("id", "1041", "item", "USB-C хаб", "status_label", "доставлен",
                        "amount", 34.5, "currency", "EUR", "url", "https://shop.test/orders/1041")));
        assertTrue(texts(msgs).contains("[Открыть карточку заказа →](https://shop.test/orders/1041)"),
                "components: " + texts(msgs));
    }

    @Test
    void orderListRowLinksTheNumber() {
        List<Map<String, Object>> msgs = A2ui.fromWidget(Map.of(
                "_kind", "widget/order_list",
                "title", "Последние заказы: alice",
                "orders", List.of(Map.of("id", "1041", "item", "USB-C хаб", "status_label", "доставлен",
                        "amount", 34.5, "currency", "EUR", "created", "2026-06-10",
                        "url", "https://shop.test/orders/1041"))));
        assertTrue(texts(msgs).contains(
                        "[#1041](https://shop.test/orders/1041)  USB-C хаб — доставлен (34.5 EUR, 2026-06-10)"),
                "components: " + texts(msgs));
    }

    @Test
    void orderWidgetWithoutUrlHasNoLink() {
        List<Map<String, Object>> msgs = A2ui.fromWidget(Map.of(
                "_kind", "widget/order",
                "title", "Заказ 1041",
                "order", Map.of("id", "1041", "item", "USB-C хаб")));
        assertTrue(texts(msgs).stream().noneMatch(s -> s.startsWith("[")),
                "no link expected: " + texts(msgs));
    }

    @Test
    void parseAction() {
        A2ui.Action a = A2ui.parseAction(Map.of(
                "version", "v0.9",
                "action", Map.of("name", "approve_refund", "context", Map.of("order_id", "1055"))));
        assertNotNull(a);
        assertEquals("approve_refund", a.name());
        assertEquals("1055", a.context().get("order_id"));
    }

    @Test
    void parseActionRejectsNonAction() {
        assertNull(A2ui.parseAction(Map.of("useStreaming", false)));
        assertNull(A2ui.parseAction(null));
        assertNull(A2ui.parseAction("not a map"));
    }
}
